sklearn/impute/_base.py

```diff
@@ -481,22 +476,15 @@ def _sparse_fit(self, X, strategy, missing_values, fill_value):
         statistics = np.empty(X.shape[1])
 
         if strategy == "constant":
-            # TODO(1.8): Remove FutureWarning and add `np.nan` as a statistic
-            # for empty features to drop them later.
-            if not self.keep_empty_features and any(
-                [all(missing_mask[:, i].data) for i in range(missing_mask.shape[1])]
-            ):
-                warnings.warn(
-                    "Currently, when `keep_empty_feature=False` and "
-                    '`strategy="constant"`, empty features are not dropped. '
-                    "This behaviour will change in version 1.8. Set "
-                    "`keep_empty_feature=True` to preserve this behaviour.",
-                    FutureWarning,
-                )
-
             # for constant strategy, self.statistics_ is used to store
             # fill_value in each column
             statistics.fill(fill_value)
+
+            if not self.keep_empty_features:
+                for i in range(missing_mask.shape[1]):
+                    if all(missing_mask[:, i].data):
+                        statistics[i] = np.nan
```

Comment:
The reason this works, but the dense case doesn't, is that here `statistics` is created as a float array, whereas there the array is created with the same dtype as the input, and then putting `np.nan` (which is a float) into the int array fails.

I think we should have probably a dtype object for statistics, and then convert the type before putting it into X when modifying it instead.